lib/valueflow.cpp

```diff
@@ -1796,7 +1796,7 @@ static std::vector<ValueFlow::LifetimeToken> getLifetimeTokens(const Token* tok,
         if (Token::Match(tok->tokAt(-2), ". %name% (") && tok->tokAt(-2)->originalName() != "->" && astIsContainer(tok->tokAt(-2)->astOperand1())) {
             const Library::Container* library = getLibraryContainer(tok->tokAt(-2)->astOperand1());
             const Library::Container::Yield y = library->getYield(tok->strAt(-1));
-            if (y == Library::Container::Yield::AT_INDEX || y == Library::Container::Yield::ITEM) {
+            if (contains({Library::Container::Yield::AT_INDEX, Library::Container::Yield::ITEM, Library::Container::Yield::BUFFER, Library::Container::Yield::BUFFER_NT}, y)) {
```

Comment:
This is not correct, The buffer returns a pointer not a reference so it should not be used here.

Reply:
Its get the lifetime of a reference. Thats why the it only follows the function that return references(see [here](https://github.com/danmar/cppcheck/blob/5f301d3285bbfbc3803eb6503de30c9b0eacd1ee/lib/valueflow.cpp#L1757)). Its similar to `followReferences`, but different as it will get the parent lifetime of a reference as well.

Originally, it was called `getLifetimeVariable`(and there is still an overload for this), but was renamed to `getLifetimeTokens` so it can handle expressions and multiple lifetimes.

The lifetime of pointers are set in valueflow in the token. For pointer to containers `data()` method, we set the lifetime here: https://github.com/danmar/cppcheck/blob/5f301d3285bbfbc3803eb6503de30c9b0eacd1ee/lib/valueflow.cpp#L3237

Reply:
Sorry, that is my fault entirely since I wrote and named this function originally.